mutable_set/mutable_set.mbt

```diff
@@ -0,0 +1,565 @@
+pub fn intersection[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T]
+) -> MutableSet[T] {
+  let res = MutableSet::with_capacity(@math.minimum(self.size(), other.size()))
+  other.iter(fn(key : T) -> Unit { if self.exists(key) { res.insert(key) } })
+  res
+}
+
+/// Returns the intersection of self with other, but store the result in `into`
+/// 
+/// # Example
+/// 
+/// ```
+/// let res = MutableSet::[]
+/// MutableSet::[3, 4, 5].inter_into(MutableSet::[4, 5, 6], res)
+/// println(res.to_array()) // output: [5, 4]
+/// ```
+pub fn intersection_into[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T],
+  into : MutableSet[T]
+) -> Unit {
+  other.iter(fn(key : T) -> Unit { if self.exists(key) { into.insert(key) } })
+}
+
+/// Returns the union of self and other.
+/// 
+/// # Example
+/// 
+/// ```
+/// let res = MutableSet::[3, 4, 5].union(MutableSet::[4, 5, 6]).to_array()
+/// res.sort()
+/// println(res) // output: [3, 4, 5, 6]
+/// ```
+pub fn union[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T]
+) -> MutableSet[T] {
+  let res = self.copy()
+  other.iter(fn(key : T) -> Unit { res.insert(key) })
+  res
+}
+
+/// Returns the union of self and other, but store the result in `into`
+/// 
+/// # Example
+/// 
+/// ```
+/// let into = MutableSet::[]
+/// MutableSet::[3, 4, 5].union_into(MutableSet::[4, 5, 6], into)
+/// let into = into.to_array()
+/// into.sort()
+/// println(res) //output: [3, 4, 5, 6]
+/// ```
+pub fn union_into[T : Hash + Eq + Show](
+  self : MutableSet[T],
+  other : MutableSet[T],
+  into : MutableSet[T]
+) -> Unit {
+  self.union(other).iter(fn(key : T) -> Unit { into.insert(key) })
+}
+
+/// Returns the difference between self and other.
+/// 
+/// # Example
+/// 
+/// ```
+/// println(MutableSet::[1, 2, 3].diff(MutableSet::[4, 5, 1]).to_array())
+/// // output: [3, 2]
+/// ```
+pub fn diff[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T]
+) -> MutableSet[T] {
+  let res = self.copy()
+  other.iter(fn(key : T) -> Unit { res.remove_without_return(key) })
+  res
+}
+
+/// Returns the difference between self and other, but store the result in `into`
+/// 
+/// # Example
+/// 
+/// ```
+/// let into = MutableSet::[]
+/// MutableSet::[1, 2, 3].diff_into(MutableSet::[4, 5, 1], into)
+/// println(into.to_array()) // output: [2, 3]
+/// ```
+pub fn diff_into[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T],
+  into : MutableSet[T]
+) -> Unit {
+  self.diff(other).iter(fn(key : T) -> Unit { into.insert(key) })
+}
+
+// The following are the helper functions used by the internal implementation of MutableSet:
+
+fn Item::new[T](key : T) -> Item[T] {
+  { key, probe_length: 0 }
+}
+
+fn insert_item[T : Hash + Eq](self : MutableSet[T], key : T) -> Int {
+  match self.index_of(key) {
+    Some(index) => {
+      self.directory[index].unwrap().key = key
+      index
+    }
+    None => {
+      if self.effective_size == self.directory.length() {
```

Comment:
The load factor is not **how** you resize the hashset; it's about **when** you resize the hashset

For example, if I understand correctly (correct me if I'm wrong), you decide to resize it when `self.effective_size == self.directory.length()`, then it means the load (number of items/length of array) will become 100% before you resize it. It may have some poor performances (when it is 90% full or 95% full or 99% full)

Reply:
Oh! I was wrong all along, that is, when the load factor is 0.8, the HashSet should be expanded when it reaches 80% capacity, not when it is full.

Reply:
So is it possible to simply modify it to:
```
if self.directory.length() * 80 / 100 == self.effective_size {
  ...
```

Reply:
```moonbit
test "" {
  let a = MutableSet::[0]
  a.insert(1)
}
```